### ISSUE TITLE
 Windows containers restart in vSphere cluster

### ISSUE THREAD

**Issue by ravisantoshgudimetla** [microsoft/Windows-Containers#68]
**What happened**:
Windows Containers are restarting in vSphere cluster on Windows 1909 host with the following error:

```
failed to shutdown container: container 54c04719fab4bed4e7a791b3c57e6a0ab4308c5aac3843221fcf79e96d855081 encountered an error during hcsshim::System::Shutdown: failure in a Windows system call: The remote procedure call failed and did not execute. (0x6bf): subsequent terminate failed container 54c04719fab4bed4e7a791b3c57e6a0ab4308c5aac3843221fcf79e96d855081 encountered an error during hcsshim::System::Terminate: failure in a Windows system call: The remote procedure call failed and did not execute. (0x6bf)
```

**What you expected to happen**:
Windows Containers are running normally.

**How to reproduce it (as minimally and precisely as possible)**:
Add a Windows 1909 host to existing OpenShift cluster with ovn-kubernetes

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.19.0+f5121a6
- Cloud provider or hardware configuration: vSphere 6.7
- OS (e.g: `cat /etc/os-release`): Microsoft Windows 1909 
- Kernel (e.g. `uname -a`): 10.0.18363.1082 - This has custom VXLAN patch enabled
- Install tools: OpenShift installer
- Network plugin and version (if this is a network-related bug): ovn-kubernetes with hybrid-overlay enabled
- Others:

**Comment by ravisantoshgudimetla** [comment]
On the Windows host when the containers are restarting, clone the [repo](https://github.com/microsoft/SDN/tree/master/Kubernetes/windows/debug) and run `startpacketcapture.cmd`

Network logs:[server.tar.gz](https://github.com/kubernetes/kubernetes/files/5398055/server.tar.gz)

System events log:[system.tar.gz](https://github.com/kubernetes/kubernetes/files/5398071/system.tar.gz)

**Comment by ravisantoshgudimetla** [comment]
cc @aravindhp @gmarkley-VI @marosset @ambguo @JocelynBerrendonner

**Comment by ravisantoshgudimetla** [comment]
Thank you @marosset for redirecting me to this repo.

**Comment by brasmith-ms** [comment]
Hi @ravisantoshgudimetla thank you for reporting this. We will take a look into this issue and report back with our findings.

**Comment by brasmith-ms** [comment]
Would you be able to provide any of the event logs from the Hyper-V Host Compute Service (HCS)? Event type: Microsoft-Windows-HyperV-Compute-Operational

**Comment by ravisantoshgudimetla** [comment]
@brasmith-ms - Thank you for responding back. Is the following command right to get the logs you requested?

`Get-EventLog system -Source *Hyper-V*`

**Comment by ravisantoshgudimetla** [comment]
Used the following documentation to create hcslog.evtx:

https://docs.microsoft.com/en-us/virtualization/windowscontainers/troubleshooting

[hcslog.tar.gz](https://github.com/microsoft/Windows-Containers/files/5418915/hcslog.tar.gz)

Which Jocelyn gave us. @brasmith-ms, please let us know if this is what you want.

**Comment by sebsoto** [comment]
I was able to reproduce this issue and I got the following:
```
[ssoto@ip-10-0-5-26 ~]$ ./oc get po -n default
NAME                             READY   STATUS    RESTARTS   AGE
win-webserver-7df5765678-vpzx6   1/1     Running   39         15h
[ssoto@ip-10-0-5-26 ~]$ ./oc describe po -n default
Name:         win-webserver-7df5765678-vpzx6
Namespace:    default
Priority:     0
Node:         winworker-tgk4d/172.31.251.167
Start Time:   Mon, 26 Oct 2020 22:03:40 +0000
Labels:       app=win-webserver
              pod-template-hash=7df5765678
Annotations:  <none>
Status:       Running
IP:           10.132.2.48
IPs:
  IP:           10.132.2.48
Controlled By:  ReplicaSet/win-webserver-7df5765678
Containers:
  windowswebserver:
    Container ID:  docker://be362ce11a522fabbfa3c474b25ae4040877c89ce9271bf840b39071a9b2db8e
    Image:         mcr.microsoft.com/powershell:lts-nanoserver-1909
    Image ID:      docker-pullable://mcr.microsoft.com/powershell@sha256:8067cc3b119547b148433a8fb934cf1407669fe61ad81364ecb68b59298c800d
    Port:          <none>
    Host Port:     <none>
    Command:
      pwsh.exe
      -command
      $listener = New-Object System.Net.HttpListener; $listener.Prefixes.Add('http://*:80/'); $listener.Start();Write-Host('Listening at http://*:80/'); while ($listener.IsListening) { $context = $listener.GetContext(); $response = $context.Response; $content='<html><body><H1>Windows Container Web Server</H1></body></html>'; $buffer = [System.Text.Encoding]::UTF8.GetBytes($content); $response.ContentLength64 = $buffer.Length; $response.OutputStream.Write($buffer, 0, $buffer.Length); $response.Close(); };
    State:          Running
      Started:      Tue, 27 Oct 2020 13:27:32 +0000
    Last State:     Terminated
      Reason:       Error
      Message:      failed to shutdown container: container 77994130cd7324f4f38626e2059ad81bb79a8fb953300a8ae7b434eb379aa6a9 encountered an error during hcsshim::System::Shutdown: failure in a Windows system call: The remote procedure call failed and did not execute. (0x6bf): subsequent terminate failed container 77994130cd7324f4f38626e2059ad81bb79a8fb953300a8ae7b434eb379aa6a9 encountered an error during hcsshim::System::Terminate: failure in a Windows system call: The remote procedure call failed and did not execute. (0x6bf)
      Exit Code:    -1
      Started:      Tue, 27 Oct 2020 13:27:01 +0000
      Finished:     Tue, 27 Oct 2020 13:27:15 +0000
    Ready:          True
    Restart Count:  39
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-22jj6 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  default-token-22jj6:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-22jj6
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  beta.kubernetes.io/os=windows
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
                 os=Windows
Events:
  Type     Reason          Age                 From     Message
  ----     ------          ----                ----     -------
  Normal   Killing         20m (x8 over 8h)    kubelet  Stopping container windowswebserver
  Normal   SandboxChanged  20m (x41 over 15h)  kubelet  Pod sandbox changed, it will be killed and re-created.
  Warning  BackOff         20m (x26 over 8h)   kubelet  Back-off restarting failed container
  Normal   Pulled          20m (x40 over 15h)  kubelet  Container image "mcr.microsoft.com/powershell:lts-nanoserver-1909" already present on machine
  Normal   Created         20m (x40 over 15h)  kubelet  Created container windowswebserver
  Normal   Started         20m (x40 over 15h)  kubelet  Started container windowswebserver

```
These restarts occurred on their own, just from the pod being deployed.

**Comment by sebsoto** [comment]
I was also able to reproduce this issue just through Docker on the same vSphere node. After stopping the running pod:
```
administrator@WINWORKER-TGK4D C:\Users\Administrator>docker run -d -p 80:80 -u ContainerAdministrator mcr.microsoft.com/powershell:lts-nanoserver-1909 pwsh.exe -command "$listener = New-Object System.Net.HttpListener; $listener.Prefixes.A
dd('http://*:80/'); $listener.Start();Write-Host('Listening at http://*:80/'); while ($listener.IsListening) { $context = $listener.GetContext(); $response = $context.Response; $content='<html><body><H1>Windows Container Web Server</H1></
body></html>'; $buffer = [System.Text.Encoding]::UTF8.GetBytes($content); $response.ContentLength64 = $buffer.Length; $response.OutputStream.Write($buffer, 0, $buffer.Length); $response.Close(); };
cbebb82ecf96e52c49b729460df493589f8952216a1d978cbeb5296ac07dae17

administrator@WINWORKER-TGK4D C:\Users\Administrator>docker ps
CONTAINER ID        IMAGE                                              COMMAND                  CREATED             STATUS              PORTS                NAMES
cbebb82ecf96        mcr.microsoft.com/powershell:lts-nanoserver-1909   "pwsh.exe -command '…"   4 seconds ago       Up 3 seconds        0.0.0.0:80->80/tcp   affectionate_gould

```
Then after waiting a period of time:
```
administrator@WINWORKER-TGK4D C:\Users\Administrator>docker ps -a
CONTAINER ID        IMAGE                                              COMMAND                  CREATED             STATUS                              PORTS               NAMES
cbebb82ecf96        mcr.microsoft.com/powershell:lts-nanoserver-1909   "pwsh.exe -command '…"   29 minutes ago      Exited (4294967295) 3 minutes ago                       affectionate_gould

administrator@WINWORKER-TGK4D C:\Users\Administrator>docker logs cbebb82ecf96
Listening at http://*:80/

administrator@WINWORKER-TGK4D C:\Users\Administrator>docker inspect cbebb82ecf96
[
    {
        "Id": "cbebb82ecf96e52c49b729460df493589f8952216a1d978cbeb5296ac07dae17",
        "Created": "2020-10-27T14:00:59.9826323Z",
        "Path": "pwsh.exe",
        "Args": [
            "-command",
            "$listener = New-Object System.Net.HttpListener; $listener.Prefixes.Add('http://*:80/'); $listener.Start();Write-Host('Listening at http://*:80/'); while ($listener.IsListening) { $context = $listener.GetContext(); $response =
 $context.Response; $content='<html><body><H1>Windows Container Web Server</H1></body></html>'; $buffer = [System.Text.Encoding]::UTF8.GetBytes($content); $response.ContentLength64 = $buffer.Length; $response.OutputStream.Write($buffer, 0
, $buffer.Length); $response.Close(); };"
        ],
        "State": {
            "Status": "exited",
            "Running": false,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 0,
            "ExitCode": 4294967295,
            "Error": "failed to shutdown container: container cbebb82ecf96e52c49b729460df493589f8952216a1d978cbeb5296ac07dae17 encountered an error during hcsshim::System::Shutdown: failure in a Windows system call: The remote procedure c
all failed and did not execute. (0x6bf): subsequent terminate failed container cbebb82ecf96e52c49b729460df493589f8952216a1d978cbeb5296ac07dae17 encountered an error during hcsshim::System::Terminate: failure in a Windows system call: The 
remote procedure call failed and did not execute. (0x6bf)",
            "StartedAt": "2020-10-27T14:01:00.5459265Z",
            "FinishedAt": "2020-10-27T10:26:53.4685869-04:00"
        },
        "Image": "sha256:47e24b62747e1b4c951ffaf8a261c4754bcf73f9e6abb1009c459b9d948b3d54",
        "ResolvConfPath": "",
        "HostnamePath": "",
        "HostsPath": "",
        "LogPath": "C:\\ProgramData\\docker\\containers\\cbebb82ecf96e52c49b729460df493589f8952216a1d978cbeb5296ac07dae17\\cbebb82ecf96e52c49b729460df493589f8952216a1d978cbeb5296ac07dae17-json.log",
        "Name": "/affectionate_gould",
        "RestartCount": 0,
        "Driver": "windowsfilter",
        "Platform": "windows",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": null,
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "default",
            "PortBindings": {
                "80/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "80"
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "no",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Capabilities": null,
            "Dns": [],
            "DnsOptions": [],
            "DnsSearch": [],
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 0,
            "ConsoleSize": [
                55,
                238
            ],
            "Isolation": "process",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": [],
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": [],
            "DeviceCgroupRules": null,
            "DeviceRequests": null,
            "KernelMemory": 0,
            "KernelMemoryTCP": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": null,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": null,
            "ReadonlyPaths": null
        },
        "GraphDriver": {
            "Data": {
                "dir": "C:\\ProgramData\\docker\\windowsfilter\\cbebb82ecf96e52c49b729460df493589f8952216a1d978cbeb5296ac07dae17"
            },
            "Name": "windowsfilter"
        },
        "Mounts": [],
        "Config": {
            "Hostname": "cbebb82ecf96",
            "Domainname": "",
            "User": "ContainerAdministrator",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "80/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "ProgramFiles=C:\\Program Files",
                "LOCALAPPDATA=C:\\Users\\ContainerAdministrator\\AppData\\Local",
                "PSModuleAnalysisCachePath=\\Microsoft\\Windows\\PowerShell\\docker\\ModuleAnalysisCache",
                "PSCORE=\\PowerShell\\pwsh.exe",
                "WindowsPATH=C:\\Windows\\system32;C:\\Windows",
                "POWERSHELL_DISTRIBUTION_CHANNEL=PSDocker-NanoServer-1909",
                "POWERSHELL_TELEMETRY_OPTOUT=0"
            ],
            "Cmd": [
                "pwsh.exe",
                "-command",
                "$listener = New-Object System.Net.HttpListener; $listener.Prefixes.Add('http://*:80/'); $listener.Start();Write-Host('Listening at http://*:80/'); while ($listener.IsListening) { $context = $listener.GetContext(); $respon
se = $context.Response; $content='<html><body><H1>Windows Container Web Server</H1></body></html>'; $buffer = [System.Text.Encoding]::UTF8.GetBytes($content); $response.ContentLength64 = $buffer.Length; $response.OutputStream.Write($buffe
r, 0, $buffer.Length); $response.Close(); };"
            ],
            "Image": "mcr.microsoft.com/powershell:lts-nanoserver-1909",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {}
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "cbebb82ecf96e52c49b729460df493589f8952216a1d978cbeb5296ac07dae17",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "cbebb82ecf96e52c49b729460df493589f8952216a1d978cbeb5296ac07dae17",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "nat": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": null,
                    "NetworkID": "e8ce3d1d9f210fdcec192a1ec6b9011320aae51626683bf0aad3fbcaa0ea4fa5",
                    "EndpointID": "",
                    "Gateway": "",
                    "IPAddress": "",
                    "IPPrefixLen": 0,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "",
                    "DriverOpts": null
                }
            }
        }
    }
]

```

**Comment by brasmith-ms** [comment]
The source of this issue has been identified. There was a plugin for VMware Tools called AppInfo that attempts to collect information from the container processes every 30 minutes. Based on the open source [header file](https://github.com/vmware/open-vm-tools/blob/f72e314e8b0df4e80c6b5f9b0c66ad2e9ce02e19/open-vm-tools/lib/include/procMgr.h#L62) and [linux implementation](https://github.com/vmware/open-vm-tools/blob/f72e314e8b0df4e80c6b5f9b0c66ad2e9ce02e19/open-vm-tools/lib/procMgr/procMgrPosix.c#L352), information is collected by walking all processes in the system and attempting to retrieve their command line. The Windows implementation for this plugin is not open source, and thus we assume there is a bug in the implementation of this plugin for Windows containers. 

This plugin can be disabled by following [these instructions](https://docs.vmware.com/en/VMware-Tools/11.0.0/com.vmware.vsphere.vmwaretools.doc/GUID-3A8089F6-CAF6-43B9-BD9D-B1081F8D64E2.html). If you have any additional concerns feel free to reach out to me.

**Comment by aravindhp** [comment]
@michmike you might someone from VMware to take a look at https://github.com/microsoft/Windows-Containers/issues/68#issuecomment-718233792

**Comment by rosskirkpat** [comment]
The issue with the appInfo plugin was fixed as of[VMWare Tools v11.0.6](https://docs.vmware.com/en/VMware-Tools/11.0/rn/VMware-Tools-1106-Release-Notes.html). 

VMWare Tools v11.0.0 to v11.0.5 have this bug present. In my case, every 30 minutes all docker containers on the host would restart while running 11.0.1

> Resolved Issues
In the 64-bit Windows guest VM, some 64-bit applications can crash (exit abruptly), due to which the corresponding services stop working. 
In VMware Tools 11.0.0, a new plugin called 'appInfo' was bundled. This new plugin runs every 30 minutes (can be configured by the guest administrator) to capture the information about running processes. For this, the plugin creates a remote thread in the target 64-bit process to capture the "command line" of the target process. If the 64-bit application of the process is not built with /LARGEADDRESSAWARE:YES, the "remote thread" crashes the application. For more information, refer KB 78550.

> This issue is fixed in this release.
